File: runtime/go/units/rt/i32_trunc_s.go
// requires: rt/trap
func (rt) i32_trunc_s(x float64) uint32 {
    if math.IsNaN(x) {
        Rt.trap("invalid conversion to integer")
    }
    if math.IsInf(x, 0) {
        Rt.trap("integer overflow")
    }
    t := math.Trunc(x)
    if t < -2147483648 || t > 2147483647 {
        Rt.trap("integer overflow")
    }
    return uint32(int32(t))
}

File: runtime/java/units/rt/i32_trunc_s.java
// requires: rt/trap
// i32.trunc_f32_s / i32.trunc_f64_s: trapping signed truncation. Java's `(int)`
// cast saturates, so wasm's two trap conditions (NaN, out-of-range) need
// explicit checks. The source is always widened to double first
// (exact for f32), so one helper serves both widths.
static int i32_trunc_s(double x) {
    if (Double.isNaN(x)) {
        trap("invalid conversion to integer");
    }
    if (Double.isInfinite(x)) {
        trap("integer overflow");
    }
    double t = x < 0 ? Math.ceil(x) : Math.floor(x);
    if (t < -2147483648.0 || t > 2147483647.0) {
        trap("integer overflow");
    }
    return (int) t;
}
